tests/components/matter/test_update.py

```diff
@@ -169,3 +179,131 @@ async def test_update_install(
     state = hass.states.get("update.mock_dimmable_light")
     assert state.state == STATE_OFF
     assert state.attributes.get("installed_version") == "v2.0"
+
+
+async def test_update_install_failure(
+    hass: HomeAssistant,
+    matter_client: MagicMock,
+    check_node_update: AsyncMock,
+    update_node: AsyncMock,
+    updateable_node: MatterNode,
+) -> None:
+    """Test update entity exists and update check got made."""
+    state = hass.states.get("update.mock_dimmable_light")
+    assert state
+    assert state.state == STATE_OFF
+    assert state.attributes.get("installed_version") == "v1.0"
+
+    await async_setup_component(hass, "homeassistant", {})
+
+    check_node_update.return_value = MatterSoftwareVersion(
+        vid=65521,
+        pid=32768,
+        software_version=2,
+        software_version_string="v2.0",
+        firmware_information="",
+        min_applicable_software_version=0,
+        max_applicable_software_version=1,
+        release_notes_url="http://home-assistant.io/non-existing-product",
+        update_source=UpdateSource.LOCAL,
+    )
+
+    await hass.services.async_call(
+        "homeassistant",
+        "update_entity",
+        {
+            ATTR_ENTITY_ID: "update.mock_dimmable_light",
+        },
+        blocking=True,
+    )
+
+    assert matter_client.check_node_update.call_count == 2
+
+    state = hass.states.get("update.mock_dimmable_light")
+    assert state
+    assert state.state == STATE_ON
+    assert state.attributes.get("latest_version") == "v2.0"
+    assert (
+        state.attributes.get("release_url")
+        == "http://home-assistant.io/non-existing-product"
+    )
+
+    update_node.side_effect = UpdateCheckError("Error finding applicable update")
+
+    with pytest.raises(HomeAssistantError):
+        await hass.services.async_call(
+            "update",
+            "install",
+            {
+                ATTR_ENTITY_ID: "update.mock_dimmable_light",
+                "version": "v3.0",
+            },
+            blocking=True,
+        )
+
+    update_node.side_effect = UpdateError("Error updating node")
+
+    with pytest.raises(HomeAssistantError):
+        await hass.services.async_call(
+            "update",
+            "install",
+            {
+                ATTR_ENTITY_ID: "update.mock_dimmable_light",
+                "version": "v3.0",
+            },
+            blocking=True,
+        )
+
+
+async def test_update_state_restore(
```

Comment:
We have `mock_restore_state_with_extra_data` to help test loading of restored state.

Example:
https://github.com/home-assistant/core/blob/7348a1fd0c9c11c3ca1a75f0c1f2e186609ec2de/tests/components/zwave_js/test_update.py#L706-L739

Reply:
I took inspiration from `test_restore_sensor_save_state` for this test.

From what I can tell `mock_restore_cache_with_extra_data` would only allow to test the restore side, but not really the saving side (making sure the update information makes it into the extra data).

The way I wrote the test is really what I (the user) is interested in: Make sure the update information survives a restart.

Reply:
I didn't mean that we should remove this test. But we need to add another test that uses that helper and tests the loading.

Reply:
I see, yeah seems sensible to test the loading side only, to make sure this remains compatible/correct.

We don't use much of the extra data currently, essentially we just use it to get the software update integer version to pass it to the update call. I guess it could be debated if the implementation of the Matter update entity should make use more of the extra data directly :thinking:  But that would need an update entity change too.

Anyways, the current test really tests what extra data is used now, so IMHO this is already a good step forward in test coverage.